File: packages/sdk/examples/exampleBrlOnramp.ts
import { CreateQuoteRequest, EPaymentMethod, EvmToken, FiatToken, Networks, QuoteResponse, RampDirection } from "../src/index";
import { VortexSdkConfig } from "../src/types";
import { VortexSdk } from "../src/VortexSdk";

function requireEnv(name: string): string {
  const value = process.env[name];
  if (!value) {
    throw new Error(`Missing required env var: ${name}. Copy .env.example to .env and set it.`);
  }
  return value;
}

async function runBrlOnrampExample() {
  try {
    console.log("Starting BRL Onramp Example...\n");

    console.log("📝 Step 1: Initializing VortexSdk...");
    const config: VortexSdkConfig = {
      apiBaseUrl: process.env.VORTEX_API_URL ?? "http://localhost:3000",
      autoReconnect: true,
      // Optional: provide custom WebSocket URLs
      moonbeamWsUrl: undefined,
      pendulumWsUrl: undefined, // 'wss://custom-moonbeam-rpc.com',
      publicKey: requireEnv("VORTEX_PUBLIC_KEY"), // 'wss://custom-pendulum-rpc.com',
      secretKey: requireEnv("VORTEX_SECRET_KEY"), // default is `true`
      // Optional: store ephemeral keys for later use
      storeEphemeralKeys: true // default is `true`
    };

    const sdk = new VortexSdk(config);

    console.log("⏳ Waiting for API initialization...");
    console.log("✅ VortexSdk initialized successfully\n");

    console.log("📝 Step 2: Creating quote for BRL onramp...");
    const quoteRequest: CreateQuoteRequest = {
      from: EPaymentMethod.PIX,
      inputAmount: "100",
      inputCurrency: FiatToken.BRL,
      network: Networks.Polygon,
      outputCurrency: EvmToken.USDC,
      rampType: RampDirection.BUY,
      to: Networks.Polygon
      //partnerId: "example-partner"
    };

    const quote = (await sdk.createQuote(quoteRequest)) as QuoteResponse;
    console.log("✅ Quote created successfully:");
    console.log(`   Quote ID: ${quote.id}`);
    console.log(`   Input: ${quote.inputAmount} ${quote.inputCurrency}`);
    console.log(`   Output: ${quote.outputAmount} ${quote.outputCurrency}`);
    console.log(`   Total Fee: ${quote.totalFeeFiat} ${quote.feeCurrency}`);
    console.log(`   Expires at: ${quote.expiresAt}\n`);

    const brlOnrampData = {
      destinationAddress: requireEnv("DESTINATION_ADDRESS"),
      taxId: requireEnv("BRL_TAX_ID")
    };

    const { rampProcess } = await sdk.registerRamp(quote, brlOnrampData);

    console.log("✅ BRL Onramp registered successfully:");
    console.log(`   Ramp ID: ${rampProcess.id}`);

    if (rampProcess.depositQrCode) {
      console.log(`   Deposit QR Code: ${rampProcess.depositQrCode}`);
    }

    // Step 4: Start the BRL onramp process AFTER PAYMENT
    console.log("📝 Step 4: Starting BRL onramp...");

    // Ensure making the payment BEFORE starting the ramp
    const _startedRamp = await sdk.startRamp(rampProcess.id);
  } catch (error) {
    console.error("❌ Error in BRL Onramp Example:", error);

    if (error instanceof Error) {
      console.error("Error message:", error.message);
      console.error("Error stack:", error.stack);
    }

    process.exit(1);
  }
}

// Run the example if this file is executed directly
if (require.main === module) {
  runBrlOnrampExample()
    .then(() => {
      console.log("\n✨ Example execution completed");
      process.exit(0);
    })
    .catch(error => {
      console.error("\n💥 Example execution failed:", error);
      process.exit(1);
    });
}

File: python-sdk/examples/example_brl_onramp.py
"""
Example: Brazilian Real (BRL) Onramp

This example demonstrates how to create a BRL onramp (PIX to crypto) using the Vortex SDK.
"""

import sys
import os

# Add parent directory to path to import vortex_sdk
sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), '..')))

from vortex_sdk import VortexSdk, RampDirection, FiatToken, EvmToken, Networks


def run_brl_onramp_example():
    """Run the BRL onramp example."""
    try:
        print("Starting BRL Onramp Example...\n")
        
        # Step 1: Initialize VortexSdk
        print("📝 Step 1: Initializing VortexSdk...")
        sdk = VortexSdk(api_base_url="https://api-staging.vortexfinance.co")
        print("✅ VortexSdk initialized successfully\n")
        
        # Step 2: Create quote for BRL onramp
        print("📝 Step 2: Creating quote for BRL onramp...")
        quote_request = {
            "from": "pix",
            "inputAmount": "100",
            "inputCurrency": FiatToken.BRL,
            "network": Networks.POLYGON,
            "outputCurrency": EvmToken.USDC,
            "rampType": RampDirection.BUY,
            "to": Networks.POLYGON,
            # "partnerId": "example-partner"  # Optional
        }
        
        quote = sdk.create_quote(quote_request)
        print("✅ Quote created successfully:")
        print(f"   Quote ID: {quote['id']}")
        print(f"   Input: {quote['inputAmount']} {quote['inputCurrency']}")
        print(f"   Output: {quote['outputAmount']} {quote['outputCurrency']}")
        print(f"   Total Fee: {quote['totalFeeFiat']} {quote['feeCurrency']}")
        print(f"   Expires at: {quote['expiresAt']}\n")
        
        # Step 3: Register the BRL onramp
        print("📝 Step 3: Registering BRL onramp...")
        brl_onramp_data = {
            "destinationAddress": "0x1234567890123456789012345678901234567890",
            "taxId": "123.456.789-00"
        }
        
        result = sdk.register_ramp(quote, brl_onramp_data)
        ramp_process = result["rampProcess"]
        
        print("✅ BRL Onramp registered successfully:")
        print(f"   Ramp ID: {ramp_process['id']}")
        
        if ramp_process.get("depositQrCode"):
            print(f"   Deposit QR Code: {ramp_process['depositQrCode']}")
        print()
        
        # Step 4: Make PIX payment
        print("🛑 IMPORTANT: Make the PIX payment NOW using the QR code above.")
        print("   The payment must be completed before starting the ramp.\n")
        
        # Wait for user confirmation (in real app, you'd verify payment another way)
        input("Press Enter after you have completed the PIX payment...")
        
        # Step 5: Start the BRL onramp process AFTER PAYMENT
        print("\n📝 Step 4: Starting BRL onramp...")
        started_ramp = sdk.start_ramp(ramp_process["id"])
        print("✅ BRL Onramp started successfully!")
        print(f"   Current Phase: {started_ramp.get('currentPhase')}")
        
    except Exception as error:
        print(f"❌ Error in BRL Onramp Example: {error}")
        if hasattr(error, '__dict__'):
            print(f"Error details: {error.__dict__}")
        sys.exit(1)


if __name__ == "__main__":
    run_brl_onramp_example()
    print("\n✨ Example execution completed")
    sys.exit(0)
